cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java

```diff
@@ -0,0 +1,550 @@
+            // store the Type of Enum for all the action pushed on stack as it will later be used to generate a
+            // method call instruction
+            actionEnumSetParamTypes.add(Type.getType(Enum.class));
+          }
+
+          // create a EnumSet from the above actions
+          invokeStatic(Type.getType(EnumSet.class),
+                       new Method("of", Type.getMethodDescriptor(Type.getType(EnumSet.class),
+                                                                 actionEnumSetParamTypes.toArray(
+                                                                   new Type[actionEnumSetParamTypes.size()]))));
+
+          // generate a call to AuthEnforceUtil#enforce with the above parameters on the stack
+          invokeStatic(AUTH_ENFORCE_UTIL_TYPE,
+                       new Method("enforce", Type.getMethodDescriptor(Type.VOID_TYPE,
+                                                                      AUTHORIZATION_ENFORCER_TYPE,
+                                                                      Type.getType(EntityId.class),
+                                                                      AUTHENTICATION_CONTEXT_TYPE,
+                                                                      Type.getType(Set.class))));
+        }
+      };
+    }
+
+    private String generateUniqueFieldName(String fieldName) {
+      // pre-pend current system time to make it unique
+      return GENERATED_FIELD_PREFIX + System.currentTimeMillis() + fieldName;
+    }
+
+    private void generateFieldAndSetter(String name, Type type) {
+      String setterMethodName = GENERATED_SETTER_METHOD_PREFIX + name;
+      // add the field
+      super.visitField(Modifier.PRIVATE, name, type.getDescriptor(), null, null);
+      // get the setter method details
+      Method method = new Method(setterMethodName, Type.getMethodDescriptor(Type.VOID_TYPE, type));
+      // add the setter method
+      MethodVisitor methodVisitor = super.visitMethod(Modifier.PRIVATE, method.getName(), method.getDescriptor(),
+                                                      null, null);
+      // Put annotation on the the setter method
+      AnnotationVisitor annotationVisitor = methodVisitor.visitAnnotation(Type.getType(Inject.class).getDescriptor(),
+                                                                          true);
+      annotationVisitor.visitEnd();
+
+      // Code generation for the setter method
+      GeneratorAdapter generatorAdapter = new GeneratorAdapter(Modifier.PRIVATE, method, methodVisitor);
+      generatorAdapter.loadThis();
+      // the setter method has only one argument which is what we want the field to be set to
+      generatorAdapter.loadArg(0);
+      generatorAdapter.putField(classType, name, type);
+      generatorAdapter.returnValue();
+      generatorAdapter.endMethod();
+    }
+
+    @Override
+    public void visitEnd() {
+      if (!methodAnnotations.isEmpty()) {
```

Comment:
This check is not necessary right? Because this 2nd pass won't get triggered if the method annotations is empty, right?